File: 198-house-robber/house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
       
        memo = {}

        def f(i):
            if i < 0:
                return 0
            if i in memo:
                return memo[i]
            memo[i]=max(f(i-1),nums[i]+f(i-2))
            return memo[i]

        return f(len(nums)-1)

File: 198-house-robber/house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int prev = 0, maximum = 0;
        
        for (int num : nums) {
            int temp = max(maximum, prev + num);
            prev = maximum;
            maximum = temp;
        }
        
        return maximum;  
    }
};